File: s21notify/queries.py
# -*- coding: utf-8 -*-
"""Тексты GraphQL-операций платформы.

ВНИМАНИЕ: GraphQL-шлюз platform.21-school.ru принимает только запросы из белого
списка — текст должен ДОСЛОВНО совпадать с тем, что шлёт официальный веб-клиент
(иначе HTTP 400 c 'REQUEST_ABSENT_IN_WHITELISTS' в заголовке x-bad-request).
Не менять форматирование и набор полей.

Тексты операций взяты из https://github.com/s21toolkit/s21schema (MIT).
"""

# ---- предстоящие проверки (я проверяю / меня проверяют) ---------------------
BOOKINGS_OP = "calendarGetMyReviews"
BOOKINGS_QUERY = """fragment ProjectTeamMember on User {
  id
  avatarUrl
  login
  userExperience {
    level {
      id
      range {
        levelCode
      }
    }
    cookiesCount
    codeReviewPoints
  }
  activeSchoolShortName
}

fragment ProjectTeamMembers on ProjectTeamMembers {
  id
  teamLead {
    ...ProjectTeamMember
  }
  members {
    ...ProjectTeamMember
  }
  invitedUsers {
    ...ProjectTeamMember
  }
  teamName
  teamStatus
  minTeamMemberCount
  maxTeamMemberCount
}

fragment UserInBooking on User {
  id
  login
  avatarUrl
  userExperience {
    level {
      id
      range {
        levelCode
      }
    }
  }
}

fragment Review on CalendarBooking {
  id
  answerId
  eventSlot {
    id
    start
    end
  }
  task {
    id
    title
    assignmentType
    goalId
    goalName
    studentTaskAdditionalAttributes {
      cookiesCount
    }
  }
  verifierUser {
    ...UserInBooking
  }
  verifiableStudent {
    id
    user {
      ...UserInBooking
    }
  }
  team {
    ...ProjectTeamMembers
  }
  bookingStatus
  isOnline
  vcLinkUrl
}

query calendarGetMyReviews($to: DateTime, $limit: Int) {
  student {
    getMyUpcomingBookings(to: $to, limit: $limit) {
      ...Review
    }
  }
}"""

# ---- лента уведомлений платформы («колокольчик») ----------------------------
NOTIFICATIONS_OP = "getUserNotifications"
NOTIFICATIONS_QUERY = """query getUserNotifications($paging: PagingInput!) {
  s21Notification {
    getS21Notifications(paging: $paging) {
      notifications {
        id
        relatedObjectType
        relatedObjectId
        message
        time
        wasRead
        groupName
      }
      totalCount
      groupNames
    }
  }
}"""

# ---- дедлайны ---------------------------------------------------------------
DEADLINES_OP = "deadlinesGetDeadlines"
DEADLINES_QUERY = """fragment Level on ExperienceLevelRange {
  id
  level
  levelCode
  leftBorder
  rightBorder
}

fragment GoalCourse on CourseCoverInformation {
  localCourseId
  courseName
  courseType
  experienceFact
  finalPercentage
  displayedCourseStatus
}

fragment DeadlineGoalData on DeadlineGoal {
  goalProjects {
    studentGoalId
    project {
      goalName
      goalId
    }
    status
    executionType
    finalPercentage
    finalPoint
    pointTask
  }
  goalCourses {
    ...GoalCourse
  }
  levels {
    ...Level
  }
}

fragment DeadlineData on Deadline {
  deadlineId
  description
  comment
  deadlineToDaysArray
  deadlineTs
  createTs
  updateTs
  status
  rules {
    logicalOperatorId
    rulesInGroup {
      logicalOperatorId
      value {
        fieldId
        subFieldKey
        operator
        value
      }
    }
  }
}

query deadlinesGetDeadlines($deadlineStatuses: [DeadlineStatus!]!, $page: PagingInput!, $deadlinesFrom: DateTime, $deadlinesTo: DateTime, $sorting: [SortingField]) {
  student {
    getDeadlines(
      deadlineStatuses: $deadlineStatuses
      page: $page
      deadlineFrom: $deadlinesFrom
      deadlineTo: $deadlinesTo
      sorting: $sorting
    ) {
      deadline {
        ...DeadlineData
      }
      shiftRequests {
        deadlineShiftRequestId
        status
        daysToShift
        createTs
      }
      deadlineGoal {
        ...DeadlineGoalData
      }
      shiftCount
    }
  }
}"""

# ---- экзамены ----------------------------------------------------------------
EXAMS_OP = "calendarGetExams"
EXAMS_QUERY = """fragment CalendarExam on Exam {
  examId
  eventId
  beginDate
  endDate
  name
  location
  maxStudentCount
  currentStudentsCount
  updateDate
  goalId
  goalName
  isWaitListActive
  isInWaitList
  stopRegisterDate
}

query calendarGetExams($from: DateTime!, $to: DateTime!) {
  student {
    getExams(from: $from, to: $to) {
      ...CalendarExam
    }
  }
}"""

# ---- агенда (мои события календаря) -------------------------------------------
AGENDA_OP = "getAgendaEvents"
AGENDA_QUERY = """query getAgendaEvents($from: DateTime!, $to: DateTime!, $limit: Int!) {
  calendarEventS21 {
    getMyAgendaEvents(from: $from, to: $to, limit: $limit) {
      agendaItemContext {
        entityId
        entityType
      }
      start
      end
      label
      description
      agendaEventType
      additionalInfo {
        key
        value
      }
    }
  }
}"""

# ---- опыт/уровень/коины --------------------------------------------------------
EXPERIENCE_OP = "getCurrentUserExperience"
EXPERIENCE_QUERY = """fragment CurrentUserExperience on UserExperience {
  id
  cookiesCount
  codeReviewPoints
  coinsCount
  level {
    id
    range {
      id
      levelCode
    }
  }
}

query getCurrentUserExperience {
  student {
    getExperience {
      ...CurrentUserExperience
    }
  }
}"""

# ---- активные P2P-заявки --------------------------------------------------------
P2P_OP = "calendarGetMyActualP2pRequests"
P2P_QUERY = """fragment CalendarP2pRequest on P2pRequest {
  p2pRequestId
  startTime
  endTime
  isOnline
  goalId
  goalName
  studentAnswerId
}

query calendarGetMyActualP2pRequests($from: DateTime!, $to: DateTime!) {
  student {
    getMyActualP2pRequests(from: $from, to: $to) {
      ...CalendarP2pRequest
    }
  }
}"""

File: crates/s21-platform/src/queries.rs
//! Тексты GraphQL-операций платформы.
//!
//! ВНИМАНИЕ: GraphQL-шлюз platform.21-school.ru принимает только запросы из
//! белого списка — текст должен ДОСЛОВНО (байт в байт) совпадать с тем, что шлёт
//! официальный веб-клиент, иначе HTTP 400 с причиной в заголовке `x-bad-request`
//! (REQUEST_ABSENT_IN_WHITELISTS). Файлы queries/*.graphql выгружены скриптом из
//! s21notify/queries.py (первоисточник — github.com/s21toolkit/s21schema, MIT);
//! .gitattributes фиксирует для них LF. Не менять форматирование и набор полей.

pub const BOOKINGS_OP: &str = "calendarGetMyReviews";
pub const BOOKINGS_QUERY: &str = include_str!("../queries/bookings.graphql");

pub const NOTIFICATIONS_OP: &str = "getUserNotifications";
pub const NOTIFICATIONS_QUERY: &str = include_str!("../queries/notifications.graphql");

pub const DEADLINES_OP: &str = "deadlinesGetDeadlines";
pub const DEADLINES_QUERY: &str = include_str!("../queries/deadlines.graphql");

pub const EXAMS_OP: &str = "calendarGetExams";
pub const EXAMS_QUERY: &str = include_str!("../queries/exams.graphql");

pub const AGENDA_OP: &str = "getAgendaEvents";
pub const AGENDA_QUERY: &str = include_str!("../queries/agenda.graphql");

pub const EXPERIENCE_OP: &str = "getCurrentUserExperience";
pub const EXPERIENCE_QUERY: &str = include_str!("../queries/experience.graphql");

#[cfg(test)]
mod tests {
    use super::*;

    /// CRLF в тексте запроса = сломанный whitelist (грабля Windows-checkout).
    #[test]
    fn тексты_без_crlf_и_без_хвостовой_пустой_строки() {
        for (name, q) in [
            ("bookings", BOOKINGS_QUERY),
            ("notifications", NOTIFICATIONS_QUERY),
            ("deadlines", DEADLINES_QUERY),
            ("exams", EXAMS_QUERY),
            ("agenda", AGENDA_QUERY),
            ("experience", EXPERIENCE_QUERY),
        ] {
            assert!(!q.contains('\r'), "{name}: CRLF в тексте запроса");
            assert!(
                q.ends_with('}'),
                "{name}: лишний хвост после закрывающей скобки"
            );
        }
    }
}
